tachyon/math/finite_fields/fp12.h

```diff
@@ -28,6 +28,89 @@ class Fp12 final : public QuadraticExtensionField<Fp12<Config>> {
   constexpr static uint64_t kDegreeOverBasePrimeField = 12;
 
   static void Init() { Config::Init(); }
+
+  // CyclotomicMultiplicativeSubgroup methods
+  Fp12& FastCyclotomicSquareInPlace() {
+    // Faster Squaring in the Cyclotomic Subgroup of Sixth Degree Extensions
+    // - Robert Granger and Michael Scott
+
+    if constexpr (BasePrimeField::Config::kModulusModSixIsOne) {
+      using Fp6Ty = BaseField;
+      using Fp2Ty = typename Fp6Ty::BaseField;
+      const Fp2Ty& r0 = this->c0_.c0_;
+      const Fp2Ty& r4 = this->c0_.c1_;
+      const Fp2Ty& r3 = this->c0_.c2_;
+      const Fp2Ty& r2 = this->c1_.c0_;
+      const Fp2Ty& r1 = this->c1_.c1_;
+      const Fp2Ty& r5 = this->c1_.c2_;
+
+      // t0 + t1 * y = (z0 + z1 * y)² = a²
+      Fp2Ty tmp = r0 * r1;
+      Fp2Ty t0 = (r0 + r1) * (Fp6Ty::Config::MulByNonResidue(r1) + r0) - tmp -
```

Comment:
I just don't get where t0, t1, t2, t3, t4, t5 is coming from. Where is it coming from?

Reply:
I commented about these, But still for some parts, i don't fully understand, either.